src/cace/amp/ion_bp.c

```diff
@@ -0,0 +1,246 @@
+ */
+
+/** @file
+ * @ingroup amp
+ * Provide an ION BP adapter for AMP messaging.
+ */
+#include "ion_bp.h"
+#include "msg.h"
+#include "cace/util/logging.h"
+#include "cace/util/defs.h"
+
+#if defined(ION_FOUND)
+#include <bp.h>
+
+void cace_amp_ion_bp_state_init(cace_amp_ion_bp_state_t *state)
+{
+    CHKVOID(state);
+    m_string_init(state->eid);
+}
+
+void cace_amp_ion_bp_state_deinit(cace_amp_ion_bp_state_t *state)
+{
+    CHKVOID(state);
+    cace_amp_ion_bp_state_unbind(state);
+    m_string_clear(state->eid);
+}
+
+int cace_amp_ion_bp_state_bind(cace_amp_ion_bp_state_t *state, const m_string_t eid)
+{
+    CHKVOID(state);
+    int res = bp_open((char *)m_string_get_cstr(eid), &(state->sap));
+    if (!res)
+    {
+        m_string_set(state->eid, eid);
+    }
+    return res;
+}
+
+void cace_amp_ion_bp_state_unbind(cace_amp_ion_bp_state_t *state)
+{
+    CHKVOID(state);
+    if (!m_string_empty_p(state->eid))
+    {
+        bp_close(state->sap);
+    }
+    m_string_reset(state->eid);
+}
+
+int cace_amp_ion_bp_send(const ari_list_t data, const cace_amm_msg_if_metadata_t *meta, void *ctx)
+{
+    CHKERR1(data);
+    CHKERR1(meta);
+    cace_amp_ion_bp_state_t *state = ctx;
+    CHKERR1(state);
+
+    Sdr sdr = bp_get_sdr();
+    if (sdr_begin_xn(sdr) < 0)
+    {
+        CACE_LOG_ERR("Unable to start transaction");
+        return 2;
+    }
+
+    int retval = 0;
+    CACE_LOG_DEBUG("Sending message with %d ARIs", ari_list_size(data));
+
+    // FIXME a copy could be avoided if this encoder wrote to SDR directly
+    m_bstring_t msgbuf;
+    m_bstring_init(msgbuf);
+    if (cace_amp_msg_encode(msgbuf, data))
+    {
+        retval = 3;
+    }
+    const size_t   msg_size  = m_bstring_size(msgbuf);
+    const uint8_t *msg_begin = m_bstring_view(msgbuf, 0, msg_size);
+
+    Object extent;
+    if (!retval)
+    {
+        extent = sdr_malloc(sdr, msg_size);
+        if (extent)
+        {
+            sdr_write(sdr, extent, (char *)msg_begin, msg_size);
+        }
+        else
+        {
+            CACE_LOG_ERR("Can't write to SDR object");
+            retval = 4;
+        }
+    }
+    if (sdr_end_xn(sdr) < 0)
+    {
+        CACE_LOG_ERR("Can't close transaction");
+    }
+
+    Object content;
+    if (!retval)
+    {
+        content = ionCreateZco(ZcoSdrSource, extent, 0, msg_size, BP_STD_PRIORITY, 0, ZcoOutbound, NULL);
+        if (content == 0 || content == (Object)ERROR)
+        {
+            CACE_LOG_ERR("Zero-Copy Object creation failed");
+            retval = 5;
+        }
+    }
+
+    if (!retval)
+    {
+        Object newBundle = 0;
+
+        CACE_LOG_DEBUG("sending datagram with %zd octets to %s", msg_size, m_string_get_cstr(meta->dest));
+        int res = bp_send(state->sap,
+                          (char *)m_string_get_cstr(meta->dest), // recipient
+                          NULL,                                  // report-to
+                          300,                                   // lifespan (?)
```

Comment:
Code itself looks great here, good use of comments.

I am curious though, is the question mark for lifespan due to the value used? Does that need to be revisited?

Reply:
I think the ION calls were carried over from the earlier implementation along with these comments. That is the "lifetime in seconds" field and I will update the comment.